refactor: Restructure project to a more conventional Android architecture

wingio opened the pull request:
- [x] Use build logic to enforce configuration conventions
	- [x] App
	- [x] Core libraries
	- [x] Compose libraries
	- [x] Feature modules
- [ ] Migrate many components to isolated core modules
	- [ ] Common (constants, common utilities, etc.)
	- [ ] Data (Repositories for the GraphQL API, REST API, and user prefs)
      - [x] GraphQL Repositories
      - [ ] Rest Repository
      - [ ] Auth repository
      - [ ] Preference Repository
      - [ ] Dialog Repository
	- [ ] Design system (Common or generic UI components)
	- [ ] Domain (For now just use cases)
	- [x] GraphQL
	- [x] Icons
	- [ ] Resources (Mostly strings for common GitHub entities or actions)
	- [ ] REST
	- [ ] UI (UI-related utilities or more complex components)
- [ ] Migrate screens to feature modules
	- [ ] Auth
	- [ ] Explore
	- [ ] File Explorer
	- [ ] Home
	- [ ] Inbox
	- [ ] Profile
	- [ ] Release
	- [ ] Repository
	- [ ] Settings

wingio committed:
Add build-logic with plugins for common configurations

wingio committed:
Move GQL api to core module

wingio committed:
Add repositories in the core data module

zt64 commented on .github/workflows/gqlsync.yml:
```diff
@@ -29,7 +29,8 @@ jobs:
       - name: Download Schema via Introspection
         run: |
-          chmod +x ./gradlew && ./gradlew :api:downloadServiceApolloSchemaFromIntrospection
+          chmod +x ./gradlew
```
You dont need to run `chmod +x gradlew`. It already has executable flag set

zt64 commented on build-logic/convention/.gitignore:
```diff
@@ -0,0 +1 @@
+/build
```
No need to have this .gitignore file, root .gitignore already takes care of `/build`

zt64 commented on build-logic/convention/build.gradle.kts:
```diff
@@ -0,0 +1,57 @@
+    compilerOptions {
+        jvmTarget = JvmTarget.JVM_17
+    }
+}
```
No need for all this, you can just use this now. It also sets android JVM so you dont have to set it there either
```kt
kotlin {
    jvmToolchain(17)
}
```

zt64 commented on build-logic/convention/src/main/kotlin/dev/materii/gloom/KotlinAndroid.kt:
```diff
@@ -0,0 +1,50 @@
+
+        compileOptions {
+            sourceCompatibility = JavaVersion.VERSION_11
+            targetCompatibility = JavaVersion.VERSION_11
```
No need for this with `jvmToolchain` call mentioned in earlier comment

zt64 commented on build-logic/convention/src/main/kotlin/dev/materii/gloom/KotlinAndroid.kt:
```diff
@@ -0,0 +1,50 @@
+        is KotlinAndroidProjectExtension -> compilerOptions
+        else -> TODO("Unsupported Kotlin project extension: $this (${this::class})")
+    }.apply {
+        jvmTarget = JvmTarget.JVM_11
```
No need for this with `jvmToolchain` call mentioned in earlier comment

zt64 commented on core/data/src/main/kotlin/dev/materii/gloom/core/data/di/RepositoryModule.kt:
```diff
@@ -0,0 +1,18 @@
+import org.koin.dsl.bind
+import org.koin.dsl.module
+
+val RepositoryModule = module {
```
```kt
val repositoryModule = module {
```
PascalCase variable naming is reserved for Compose UI related variables

zt64 commented on core/graphql/src/main/kotlin/dev/materii/gloom/core/graphql/GraphQLDataSource.kt:
```diff
@@ -0,0 +1,471 @@
+        after: String?,
+        count: Int
+    ): GraphQLResponse<RepoListQuery.Data> {
+        return apolloClient.query(
+            RepoListQuery(
+                username = login,
+                cursor = after.toOptional(),
+                total = count.toOptional()
+            )
+        )
+            .addToken()
+            .response()
```
This chaining is kinda ugly I think. Looks better like this:
```kt
        return apolloClient
            .query(
                RepoListQuery(
                    username = login,
                    cursor = after.toOptional(),
                    total = count.toOptional()
                )
            )
            .addToken()
            .response()
```
Or
```kt
        return apolloClient.query(
                RepoListQuery(
                    username = login,
                    cursor = after.toOptional(),
                    total = count.toOptional()
                )
            )
            .addToken()
            .response()
```
same for the other chained functions

zt64 commented on core/graphql/src/main/kotlin/dev/materii/gloom/core/graphql/GraphQLDataSource.kt:
```diff
@@ -0,0 +1,471 @@
+    }
+
+    override suspend fun identify(): GraphQLResponse<IdentifyQuery.Data> {
+        return apolloClient.query(IdentifyQuery())
+            .addToken()
+            .response()
```
It might be better to make a function in this class which takes the query and returns the result. Cause right now its a lot of duplicate code. It could shorten this file a good bit
something like
```kt
return handleQuery(IdentityQuery())
```

zt64 commented on core/graphql/src/main/kotlin/dev/materii/gloom/core/graphql/di/GraphQLModule.kt:
```diff
@@ -0,0 +1,30 @@
+import org.koin.dsl.bind
+import org.koin.dsl.module
+
+val GraphQLModule = module {
```
```kt
val graphQLModule = module {
```

PascalCase variable naming is reserved for Compose UI related variables

zt64 commented on core/graphql/src/main/kotlin/dev/materii/gloom/core/graphql/response/GraphQLResponse.kt:
```diff
@@ -0,0 +1,31 @@
+import com.apollographql.apollo.ApolloCall
+import com.apollographql.apollo.api.Operation
+
+internal typealias GQLErrors = List<com.apollographql.apollo.api.Error>
```
Forgot to add import

zt64 commented on gradle/libs.versions.toml:
```diff
@@ -103,6 +110,12 @@ highlights = { group = "dev.snipme", name = "highlights", version.ref = "highlig
 zoomable = { group = "net.engawapg.lib", name = "zoomable", version.ref = "zoomable" }
 
+# Build-logic
+android-gradle-plugin = { group = "com.android.tools.build", name = "gradle", version.ref = "agp" }
+compose-gradle-plugin = { group = "org.jetbrains.kotlin", name = "compose-compiler-gradle-plugin", version.ref = "kotlin" }
```
You can use this function in your build-logic to directly use the plugins instead of having to redeclare them as libraries
```kt
fun DependencyHandler.compileOnly(dependency: Provider<PluginDependency>) {
    compileOnly(dependency.map { "${it.pluginId}:${it.pluginId}.gradle.plugin:${it.version}" })
}
```

zt64 commented on core/graphql/src/main/kotlin/dev/materii/gloom/core/graphql/util/Util.kt:
```diff
@@ -0,0 +1,5 @@
+
+import com.apollographql.apollo.api.Optional
+
+fun <T> T?.toOptional(): Optional<T> = Optional.presentIfNotNull(this)
```
Maybe this should be `inline`

wingio replied:
i get yelled at for it not having significant performance benefits

zt64 commented on settings.gradle.kts:
```diff
@@ -9,8 +10,19 @@ pluginManagement {
 rootProject.name = "Gloom"
 include(":app")
 
+enableFeaturePreview("TYPESAFE_PROJECT_ACCESSORS")
```
I think this should be at the very top, since it's enabling a global feature

zt64 commented on settings.gradle.kts:
```diff
@@ -9,8 +10,19 @@ pluginManagement {
 rootProject.name = "Gloom"
 include(":app")
```
`app` is included twice

wingio committed:
Move Android SDK to version catalog

 - Switch to jvmToolchain
 - Allow using existing plugin declarations from version catalog as compileOnly

wingio committed:
Move GraphQL requests to flows

wingio committed:
Add some custom models designed to improve type safety and reducing chaining

wingio committed:
Utilize new models in some api repositories

wingio committed:
Make :core:graphql use :core:common-api

Also minor fragment changes

wingio committed:
Organize and centralize icons into :core:icons

wingio committed:
Update :app gradle config to use convention plugins

wingio committed:
chore: Bump dependencies

WARNING: Only gradle sync was tested, building probably fails

wingio committed:
build(core/graphql): Correct introspection schema output path
